Add logic for reflecting IDREF (single or FrozenArray) attributes

alice opened the pull request:
Add logic for reflecting `Element?` or `FrozenArray<Element>?` attributes.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/3917/common-dom-interfaces.html" title="Last updated on Apr 6, 2021, 9:12 AM UTC (b2ce057)">/common-dom-interfaces.html</a>  ( <a href="https://whatpr.org/html/3917/e8417e5...b2ce057/common-dom-interfaces.html" title="Last updated on Apr 6, 2021, 9:12 AM UTC (b2ce057)">diff</a> )
<a href="https://whatpr.org/html/3917/infrastructure.html" title="Last updated on Apr 6, 2021, 9:12 AM UTC (b2ce057)">/infrastructure.html</a>  ( <a href="https://whatpr.org/html/3917/e8417e5...b2ce057/infrastructure.html" title="Last updated on Apr 6, 2021, 9:12 AM UTC (b2ce057)">diff</a> )

domenic commented on source:
```diff
@@ -7309,8 +7309,71 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
 
-  </div>
+  <p>If a reflecting IDL attribute has the type <code>HTMLElement</code>,
+  or an interface that descends from <code>HTMLElement</code>,
```
s/descends/inherits

alice replied:
Done.

domenic commented on source:
```diff
@@ -7309,8 +7309,71 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+  that has the type <code data-x="idl-DOMString">DOMString?</code>.
+  The relationship between these two attributes is as follows:</p>
+
+  <ol>
```
`<ul>`, not `<ol>`

alice replied:
Done.

domenic commented on source:
```diff
@@ -7309,8 +7309,71 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+  The relationship between these two attributes is as follows:</p>
+
+  <ol>
+    <li>The attribute which has the type <code>HTMLElement</code>
```
`<p>` inside all `<li>`s

alice replied:
Done.

domenic commented on source:
```diff
@@ -7309,8 +7309,71 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+
+  <ol>
+    <li>The attribute which has the type <code>HTMLElement</code>
+      should be named <code data-x=""><var>attr</var>Element</code>,
```
s/should/must

alice replied:
Done.

domenic commented on source:
```diff
@@ -7309,8 +7309,71 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+  <ol>
+    <li>The attribute which has the type <code>HTMLElement</code>
+      should be named <code data-x=""><var>attr</var>Element</code>,
+      where the paired nullable <code data-x="idl-DOMString">DOMString</code> attribute is named
```
s/is/must be/

alice replied:
Hm, I'm trying to define attrElement in terms of attr, so it's kind of tautological to say attr must be named attr, I think.

domenic commented on source:
```diff
@@ -7309,8 +7309,71 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+      where the paired nullable <code data-x="idl-DOMString">DOMString</code> attribute is named
+      <code data-x=""><var>attr</var></code>.</li>
+
+    <li>On getting <code data-x=""><var>attr</var></code> attribute,
```
Instead of "On getting _attr_ attribute, the IDL attribute must return...", say "On getting the _attr_ IDL attribute, return..."

Similar phrasing tweaks for the rest of them. It should always be "On [get|set]ing the _X_ IDL attribute, [imperative phrasing]".

domenic commented on source:
```diff
@@ -7309,8 +7309,71 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
   whose associated attribute's local name is the name of the attribute in question.</p>
 
-  </div>
+  <p>If a reflecting IDL attribute has the type <code>HTMLElement</code>,
```
I think we'll need a slightly new paradigm, since conventional reflection is more of a 1:1 relationship. Maybe something like

> In addition to simple [reflection], a pair of IDL attributes can be said to **element-reflect** a content attribute. This means that (brief summary description of the behavior, if possible).

alice replied:
Deferring this for now.

domenic commented on source:
```diff
@@ -7309,8 +7309,71 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+      the IDL attribute must return the value of the content attribute,
+      or null if the content attribute does not exist on the element.</li>
+
+    <li>On getting the <code data-x=""><var>attr</var>Element</code> IDL attribute,
```
This is not great. The better model is to say something in the intro to this element-reflect concept like "The element to which the content attribute belongs has an **_attr_-associated element**", and then operate on that concept. (Better names appreciated for that concept.)

alice replied:
Deferring this for now.

domenic commented on source:
```diff
@@ -7309,8 +7309,71 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+    </li>
+
+    <li>On setting the <code data-x=""><var>attr</var></code> attribute,
+      it must run the following algorithm:
```
You can make this a bit nicer by saying "On setting the attr attribute to _newValue_" and then using _newValue_ instead of "the specified value"

alice replied:
Done.

domenic commented on source:
```diff
@@ -7309,8 +7309,71 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+    <li>On setting the <code data-x=""><var>attr</var></code> attribute,
+      it must run the following algorithm:
+      <ol>
+        <li>If the specified value is different from the previous value of the content attribute,
```
Not previous, current.

alice replied:
Done.

domenic commented on source:
```diff
@@ -7309,8 +7309,71 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+      <ol>
+        <li>If the specified value is different from the previous value of the content attribute,
+          the stored value, if any, for <code data-x=""><var>attr</var>Element</code>
+          should be deleted.
```
With the "_attr_-associated element" concept, this can be made more precise. Probably setting it to null.

Also, instead of saying "should", you need to say "must", or better yet, just use the imperative form ("delete" or "set"). In general all the below steps should be rephrased into imperative form.

alice replied:
Rephrased as imperative; deferred the *attr*-associated element concept part.

domenic commented on source:
```diff
@@ -7309,8 +7309,71 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+        <li>If the specified value is different from the previous value of the content attribute,
+          the stored value, if any, for <code data-x=""><var>attr</var>Element</code>
+          should be deleted.
+        <li>If the specified value is null, the content attribute must be removed,
```
Use the same phrasing as below: "remove the content attribute and return"

alice replied:
Done.

domenic commented on source:
```diff
@@ -7309,8 +7309,71 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+        <li>If the specified value is null, the content attribute must be removed,
+          and the algorithm terminated.
+        <li>Otherwise, the content attribute must be set to the specified value.
+        <li>If the specified value is not a valid value for the
```
I don't think we should include this step. "valid value" would need to be defined. Probably it's better to just let the next step find nothing.

alice replied:
Done.

domenic commented on source:
```diff
@@ -7309,8 +7309,71 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+          <code data-x="attr-id">id</code> attribute,
+          the algorithm is terminated.
+        <li>Otherwise, let <var>candidate</var> be the element that the
+          <code data-x="dom-Document-getElementById">document.getElementById()</code> method
```
We'll want to rephrase this not to call the public API later. But for now it'd make things clearer to remove the `document.`, because the root may be a shadow root, in particular.

alice replied:
I reworded the text from the definition of `getElementById`.

domenic commented on source:
```diff
@@ -7309,8 +7309,71 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+          would find when called on the content attribute's element's node tree's
+          <span>root</span> if it were passed as its argument the current value
+          of the corresponding content attribute.</li>
+        <li>If <var>candidate</var> is not null, and it is type-compatible with the
```
s/it is type-compatible with the/it implements the interface given by the type of the/

alice replied:
Done.

domenic commented on source:
```diff
@@ -7309,8 +7309,71 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+    </li>
 
+    <li>On setting the <code data-x=""><var>attr</var>Element</code> attribute,
+        it must run the following algorithm:</li>
```
Again you can use _newValue_ and then use that in place of "the given element".

alice replied:
Done (used *newElement*).

domenic commented on source:
```diff
@@ -7309,8 +7309,71 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+        it must run the following algorithm:</li>
+    <ol>
+      <li>If the given element is not in the element's node tree,
+        or its light tree if it exists,
```
"light tree" is not a defined concept these days. I think the first clause is probably enough?

Here and below, it appears the canonical phrasing is "X and Y are [not] in the same tree".

alice replied:
I'm trying to say "same tree, *or* same tree as shadow host" - or maybe "same tree as any shadow-including inclusive ancestor"?

domenic commented on source:
```diff
@@ -7309,8 +7309,71 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+    <ol>
+      <li>If the given element is not in the element's node tree,
+        or its light tree if it exists,
+        or is not type-compatible with the IDL attribute,
```
type-compatibility is guaranteed here by the IDL layer, so this is redundant.

alice replied:
Done.

domenic commented on source:
```diff
@@ -7309,8 +7309,71 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+      or null if the value was not previously set or was deleted.
+    </li>
+
+    <li>On setting the <code data-x=""><var>attr</var></code> attribute,
```
"IDL" attribute here and below

domenic commented on source:
```diff
@@ -43241,23 +43305,19 @@ interface <dfn>HTMLLabelElement</dfn> : <span>HTMLElement</span> {
 
   </div>
-
-  <p>The <dfn><code data-x="attr-label-for">for</code></dfn> attribute may be specified to indicate a
```
This deletes some useful text about the for content attribute.

I think you only want to delete the part inside `<span w-nodev>`.

domenic replied:
Still, we need to have some definition of the for="" content attribute; this deletes it. And it is still accurate to say that if it is specified it must be an ID of a labelable element, as authoring guidance.

domenic commented on source:
```diff
@@ -43241,23 +43305,19 @@ interface <dfn>HTMLLabelElement</dfn> : <span>HTMLElement</span> {
-  of the <code data-x="attr-label-for">for</code> attribute, and the first such element in
-  <span>tree order</span> is a <span data-x="category-label">labelable element</span>, then that
-  element is the <code>label</code> element's <span>labeled control</span>.</span></p>
+  <p>The <code data-x="dom-label-htmlForElement">htmlForElement</code> IDL attribute
```
I think this first sentence would be better added as an edit to the sentence

> The caption can be associated with a specific form control, known as the label element's labeled control, either using the for attribute, or by putting the form control inside the label element itself.

explaining this third way of labeling things.

alice replied:
Done, I think.

domenic commented on source:
```diff
@@ -43241,23 +43305,19 @@ interface <dfn>HTMLLabelElement</dfn> : <span>HTMLElement</span> {
+  <p>The <code data-x="dom-label-htmlForElement">htmlForElement</code> IDL attribute
+  may be specified to indicate a
+  form control with which the caption is to be associated.
+  If the attribute is specified, the attribute's value must be a
```
I think we can remove this sentence. We don't normally state requirements for authors on how they use IDL attributes, especially since the type system already enforces them.

alice replied:
Done.

domenic commented on source:
```diff
@@ -43241,23 +43305,19 @@ interface <dfn>HTMLLabelElement</dfn> : <span>HTMLElement</span> {
 
   <div w-nodev>
 
-  <p>If the <code data-x="attr-label-for">for</code> attribute is not specified, but the
```
I think here we want to insert a single consolidated definition of "labeled control". So remove the sentence that says

> Except where otherwise specified by the following rules, a label element has no labeled control.

above and say something like

A label element's labeled control is determined as follows:

- (... If `label`-associated element is not null, return it ...)
- (... If descendant meeting criteria exists, return that descendant ...)
- (... there is no labeled control ...)

alice replied:
Deferred for now.

domenic commented on source:
```diff
@@ -7309,8 +7309,67 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+      where the paired nullable <code data-x="idl-DOMString">DOMString</code> attribute is named
+      <code data-x=""><var>attr</var></code>.</p></li>
 
+    <li><p>On getting <code data-x=""><var>attr</var></code> IDL attribute,
```
Missing "the"

annevk commented:
I couldn't immediately find context for the question, but I'd compare the https://dom.spec.whatwg.org/#concept-tree-root of the objects in question instead of talking about trees. There's also https://dom.spec.whatwg.org/#concept-shadow-including-root that will effectively "ignore" `ShadowRoot` as a root.

domenic commented:
Issue backreferences: https://github.com/whatwg/html/issues/3219 https://github.com/whatwg/html/issues/3515

annevk commented on source:
```diff
@@ -7369,8 +7369,111 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
 
-  </div>
+  <p>If a reflecting IDL attribute has the type <code>HTMLElement</code>,
+  or an interface that descends from <code>HTMLElement</code>, then:
```
Instead of descends, you want https://heycam.github.io/webidl/#dfn-inherit.

alice replied:
Done.

annevk commented on source:
```diff
@@ -7369,8 +7369,111 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+        <li>Let <var>candidate</var> be the element that the
+            <code data-x="dom-Document-getElementById">document.getElementById()</code> method
+            would find when called on the content attribute's element's <span>node document</span>
+            if it were passed <var>id</var> as its argument.</li>
```
I'd prefer reusing or abstracting the definition at https://dom.spec.whatwg.org/#dom-nonelementparentnode-getelementbyid. (E.g., what if someone overrode `getElementById()`?)

alice replied:
Done.

annevk commented on source:
```diff
@@ -7369,8 +7369,111 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+            would find when called on the content attribute's element's <span>node document</span>
+            if it were passed <var>id</var> as its argument.</li>
+
+        <li>If <var>candidate</var> is null, or if it is not type-compatible with the IDL attribute,
```
Ah, so this cannot return SVG elements or "XML" elements? Is that restriction really needed?

alice replied:
Hm, I replaced `HTMLElement` with `Element` above. So you could restrict it to `HTMLElement` or `SVGElement` or `HTMLInputElement` etc. if you wanted to, but the reflection mechanism doesn't require it.

annevk commented on source:
```diff
@@ -7369,8 +7369,111 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+    </li>
+    <li>On setting, it must run the following algorithm:
+      <ol>
+        <li>If the given value is null, remove the content attribute and stop.</li>
```
and return*

alice replied:
Done.

annevk commented on source:
```diff
@@ -7369,8 +7369,111 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+      <ol>
+        <li>If the given value is null, remove the content attribute and stop.</li>
+
+        <li>Otherwise, let <var>id</var> be the empty string.</li>
```
No need for "Otherwise" as you already returned.

alice replied:
I realise that's logically true, but this isn't code and we don't have indents, so I think the Otherwise... makes it easier to quickly scan and understand the algorithm.

annevk replied:
The main problem in this specific case is that you're initializing a variable that you expect to be present for all subsequent steps. I scanned this algorithm, saw that, and noticed it wasn't set in the first step, and then wondered how that would not be problematic. I think it's better to initialize variables unconditionally.

(We're also very much trying to standardize the way we write algorithms with https://infra.spec.whatwg.org/.)

annevk replied:
(To be clear, there's no need to address this nit as this isn't formalized. I mainly wanted to explain why I made the nit to begin with.)

annevk commented on source:
```diff
@@ -7369,8 +7369,111 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+        <li>Otherwise, let <var>id</var> be the empty string.</li>
+
+        <li>If the given element has an <code data-x="attr-id">id</code> attribute,
+            and has the same <span>tree</span> as the element of the attribute being set,
```
I got feedback before that it's preferable to compare the root of both elements being equal as they cannot really be said to have a tree, which I thought was fair.

alice replied:
Good suggestion, done.

annevk commented on source:
```diff
@@ -7369,8 +7369,111 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+  <ul>
+    <li>On getting, it must run the following algorithm (stopping at the first point where a value is returned):
+      <ol>
+        <li>If a value for the IDL attribute was previously set via the setting algorithm (below),
```
I think it would be better to define this through an internal slot on the element that holds the element. We're trying to stay clear of too much implicitness these days as it's not always understood.

alice replied:
Had a go at this, please let me know if you'd like it to look different.

annevk commented on source:
```diff
@@ -7369,8 +7369,111 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+        <li>Otherwise, return <var>candidate</var>.</li>
+      </ol>
+    </li>
+    <li>On setting, it must run the following algorithm:
```
It seems this setting algorithm doesn't care for the assigned element being in the same tree? Adding the internal slot to make it explicit when setting succeed would add some clarity here too.

alice replied:
Had a go at addressing this.

alice commented:
Actually updated the PR now... sorry for spam!

domenic commented on source:
```diff
@@ -7369,8 +7369,161 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
 
-  </div>
+  <p>If a reflecting IDL attribute <var>attr</var> has the type <code>Element</code>,
+  or an interface that inherits from <code>Element</code>, then:
```
The type has to be `Element?` or similar, right? If it returns or accepts `null`, then just `Element` does not suffice.

The markup would be `<code data-x=""><code>Element</code>?</code>`, it looks like, similar to what I've done below for `sequence<Element>`.

domenic commented on source:
```diff
@@ -7369,8 +7369,161 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+
+  <p>If a reflecting IDL attribute <var>attr</var> has the type <code
+  data-x="">sequence&lt;<code>Element</code>&gt;</code>, or <code data-x="">sequence&lt;<var>T</var>&gt;</code> where
+  <var>T</var> inherits from <code>Element</code>, then:</p>
```
Similarly this seems to need to be `sequence<Element>?`, right?

domenic commented on source:
```diff
@@ -7369,8 +7369,169 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+    </ol>
+   </li>
+
+   <li><p>If the content attribute is set directly (such as via <code
```
I'm not sure I understand this step. How does it impact the internal slot value? What does it mean to reset an IDL attribute's value, when an IDL attribute is actually a getter/setter as given by the above algorithms?

It seems like this step should be deleted entirely, and whatever it is trying to express, should be incorporated into the getter algorithm.

domenic replied:
Per our discussion offline, summarized in https://github.com/whatwg/html/issues/3515#issuecomment-452091002, this step is right in spirit, just needs to be formalized. The bullet should say that the element gets [attribute change steps](https://dom.spec.whatwg.org/#concept-element-attributes-change-ext) which, as their first step, check that _namespace_ is null and that _localName_ is equal to _attr_. See the example slightly below https://html.spec.whatwg.org/#event-handler-content-attributes for one current usage.

alice replied:
Ok, I had a go at this - it really does end up being exactly one step as far as I can tell (just nulling out the internal slot). PTAL.

domenic commented on source:
```diff
@@ -7369,8 +7369,169 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+
+     <li>
+      <p><span data-x="list iterate">For each</span> <var>id</var> in the <span>set of space-separated
+      tokens</span>:</p>
```
The value is not a set of space-separated tokens; it's a string. To get to a set of tokens, you need to `Let <var>tokens</var> be the content attribute's value, <span data-x="split a string on ASCII whitespace">split on ASCII whitespace</span>.`

alice commented:
Oh no: "Sequences must not be used as the type of an attribute or constant. ...  Instead of a writable attribute of a sequence type, it is suggested that a pair of operations to get and set the sequence is used."

https://heycam.github.io/webidl/#idl-sequence

domenic commented:
Ah, right, glad Web IDL caught that! We actually want to use `FrozenArray<>`, not `sequence<>`.

This will be a little annoying as we can no longer return a new list every time in the getter (i.e. "Return _elements_.") But that's actually a good fix to make. The spec right now implies that `el.attrElements !== el.attrElements`, because each call of the getter would return a fresh sequence. That's pretty bad behavior.

The most straightforward thing I can think of is to just have a separate slot, call it [[attrElementsCache]] or similar, which stores the last result of the computation logic ("On getting" steps 3-5). Then, still keep those steps to re-compute the _elements_ list, but do a comparison between the _elements_ list and the value of [[attrElementsCache]] before returning. If they're the same, return the cached value (throwing away _elements_). If they're different, reset the cached value to the new _elements_, and return it.

alice commented:
@domenic @annevk PTAL?

I reworked it in terms of `FrozenArray`. Rather than add a new slot, I have instead modified the attribute change steps in each case to compute the new value for the existing slot. Hope that's ok.

I tried using `FrozenArray<Element>` in the ARIA `AriaAttributes` IDL and the parser didn't complain, so that's a start..

annevk commented:
It makes sense for the getter to return a FrozenArray (I'm still not sure I like the freezing bit), but the setter presumably still needs to take a sequence<>? (Unfortunately not supported by IDL, a known issue of sorts.) Or would you expect users of this API to create an array, freeze it, and pass it in? Does IDL even do the necessary validation for that?

domenic commented:
When you use the `FrozenArray<>` syntax, IDL automatically converts any incoming JS iterable into an IDL frozen array. Just like when you use the `sequence<>` syntax, IDL automatically converts any incoming JS iterable into an IDL sequence.

alice commented:
Does the rest of the change look good to the two of you?

alice commented on source:
```diff
@@ -7369,8 +7370,182 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
-  </div>
+  <p>If a reflecting IDL attribute <var>attr</var> has the type <code
+  data-x=""><code>Element</code>?</code>, or an interface that inherits from <code
+  data-x=""><code>Element</code>?</code>, then:
```
Fixed, thanks.

alice commented on source:
```diff
@@ -7369,8 +7370,182 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+     <li>
+      <p>If the given value is null, or is not type-compatible with the IDL attribute, or is not a
+      descendant of either the attribute's element's <span>root</span> or the attribute's element's
+      <span>shadow-including root</span>:</p>
```
Good catch. I've tried to rewrite it so it can be in the element's tree or the host's tree - I'm not sure if we want/need to allow arbitrary levels of nesting here (e.g. host's host's host's tree).

annevk replied:
The text that there's there, "descendant of a shadow-including ancestor", would allow that...

alice replied:
Right, that was after I rewrote it again :)

alice commented on source:
```diff
@@ -7369,8 +7370,182 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+      data-x="concept-ID">ID</span> is the value of that <code data-x="attr-id">id</code>
+      attribute,</p></li>
+     </ul>
+     <p>then let <var>id</var> be the element's <span data-x="concept-ID">ID</span>.
```
Fixed, thanks.

alice commented on source:
```diff
@@ -7369,8 +7370,182 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+       <li><p>Otherwise, let <var>id</var> be the element's <code
+       data-x="concept-ID">ID</code>.</p></li>
+
+       <li><p>If <var>value</var> is non-empty, append U+0020 SPACE to <var>value</var>.</p></li>
```
Done.

alice commented on source:
```diff
@@ -7369,8 +7370,182 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+         data-x="concept-ID">ID</span> is the value of that <code data-x="attr-id">id</code>
+         attribute,</p></li>
+        </ul>
+        <p>then reset <var>value</var> to be the empty string, and continue.</p>
```
Done.

alice commented on source:
```diff
@@ -7369,8 +7370,182 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+     </ul>
+     <p>then let <var>id</var> be the element's <span data-x="concept-ID">ID</span>.
+
+     <li><p>Set the content attribute's value to <var>id</var>.</p></li>
```
Good catch - would it be worth trying to be more efficient in early-outing the attribute change steps if there will be no change?

annevk replied:
A note here acknowledging that this happens and is not a problem would be sufficient I think.

alice replied:
Hm, actually I'm not convinced triggering these steps from IDL attribute reflection is safe. 

For example (based on the [test examples](https://github.com/whatwg/html/pull/3917#issuecomment-454206464) I posted earlier):

```js
bowl.apricot = apricotOnTable;
```

This will set the content attribute to be the empty string, triggering the attribute change steps, looking for an element with an ID equal to the empty string. Since the empty string is not a valid ID, this will return null, which will set the [[apricotAssociatedElement]] slot back to null.

I added a step which early-outs if the content attribute was set via reflection of the IDL attribute. No idea if this is correct, but I think this is the idea needed to prevent weirdness.

alice commented on source:
```diff
@@ -7369,8 +7370,182 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+     <li>
+      <p>If <var>value</var> is null, then:</p>
+      <ol>
+       <li><p>Reset [[<var>attr</var>AssociatedElements]] to null.</p></li>
```
Done.

alice commented:
Some examples which may be turned into WPTs at the appropriate time:

```idl
[Reflect] attribute Element? apricot;
[Reflect] attribute FrozenArray<Element>? grapes;
```

```html
<fruit-apricot id="apricotOnFloor"></fruit-apricot>
<fruit-grape id="grapeOnFloor"></fruit-grape>
<dining-table>
  #shadow-root (open)
  | <fruit-apricot id="apricotOnTable"></fruit-apricot>
  | <fruit-grape id="grapeOnTable"></fruit-grape>
  | <centre-piece>
    #shadow-root (open)
    | <fruit-bowl id="bowl" apricot="apricot" grapes="grape1 grape2">
    |   <fruit-apricot></fruit-apricot>
    |   <fruit-apricot id="apricot"></fruit-apricot>
    |   <fruit-grape id="grape1"></fruit-grape>
    |   <fruit-grape id="grape2"></fruit-grape>    
    |   <fruit-grape></fruit-grape>
    | </fruit-bowl>
   </centre-piece>
</dining-table>
```

```js
// Let's assume we can magically get elements by their ID in any context

console.log(bowl.apricot);  // logs reference to <fruit-apricot id="apricot">
console.log(bowl.grapes);   // logs [<fruit-grape id="grape1">, <fruit-grape id="grape2">]

// ~~ No ID ~~
bowl.apricot = bowl.firstElementChild;
console.log(bowl.getAttribute("apricot"));  // logs ""

bowl.grapes = [grape1, grape2, bowl.lastElementChild];
console.log(bowl.getAttribute("grapes"));   // logs ""

// ~~ Elements in shadow host's tree ~~
bowl.apricot = apricotOnTable;
console.log(bowl.getAttribute("apricot"));  // logs ""
console.log(bowl.apricot);                  // logs reference to <fruit-apricot id="apricotOnTable"></fruit-apricot>

bowl.grapes = [grape1, grape2, grapeOnTable];
console.log(bowl.getAttribute("grapes"));   // logs ""
console.log(bowl.grapes);                   // logs [<fruit-grape id="grape1">, <fruit-grape id="grape2">, <fruit-grape id="grapeOnTable">]

// ~~ Elements in neither same tree nor shadow host's tree ~~
// Should this be an error, or fail silently? Currently fails silently.
bowl.apricot = apricotOnFloor;
console.log(bowl.getAttribute("apricot"));  // logs null
console.log(bowl.apricot);                  // logs null

bowl.grapes = [grape1, grape2, grapeOnFloor];
console.log(bowl.getAttribute("grapes"));   // logs "grape1 grape2"
console.log(bowl.grapes);                   // logs [<fruit-grape id="grape1">, <fruit-grape id="grape2">]

// ~~ Using setAttribute ~~
bowl.setAttribute("apricot", null);
console.log(bowl.apricot);                  // logs null

bowl.setAttribute("grapes", null);
console.log(bowl.grapes);                   // logs null

bowl.setAttribute("apricot", "");
console.log(bowl.apricot);                  // logs null

bowl.setAttribute("grapes", "");
console.log(bowl.grapes);                   // logs []

bowl.setAttribute("apricot", "apricot");
console.log(bowl.apricot);                  // logs reference to <fruit-apricot id="apricot">

bowl.setAttribute("grapes", "grape1 grape2");
console.log(bowl.grapes);                   // logs [<fruit-grape id="grape1">, <fruit-grape id="grape2">]

bowl.setAttribute("grapes", "grape1 grape2 grapeOnTable");
console.log(bowl.getAttribute("grapes"));   // logs "grape1 grape2 grapeOnTable"
console.log(bowl.grapes);                   // logs [<fruit-grape id="grape1">, <fruit-grape id="grape2">]
```

alice commented:
After talking with @robdodson, changed it to allow setting a reference to any element in a less-deeply-nested shadow tree.

alice commented on source:
```diff
@@ -7371,8 +7374,199 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+     <li><p>If <var>localName</var> is not associated attribute’s local name, or
+     <var>namespace</var> is not null, return.</p></li>
+
+     <li><p>If the content attribute was set via reflection of the IDL attribute, return.</p></li>
```
@domenic @annevk Does this make sense?

annevk replied:
I think it'd be better to have some kind of flag (probably in the form of an internal slot) that you set before you set the content attribute, this bails out if that flag is set, and then unset it after you've set the content attribute.

For new algorithms we try to make all state an implementation needs explicit and have less "hidden complexity", if you will.

alice replied:
Ok, I'll do that. 

I'd considered that initially but was thrown off by the idea of adding *two* internal slots for each of these things, and also that it seemed overly prescriptive when implementors may choose to implement it any number of different ways.

alice replied:
Done :)

rniwa commented on source:
```diff
@@ -7371,8 +7374,229 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+
+     <li><p>If the given element:</p>
+     <ul class="brief">
+      <li>has an <code data-x="attr-id">id</code> attribute, and</li>
```
`id` *content* attribute.

alice commented:
@cookiecrook FYI

cookiecrook commented:
@rniwa FYI

zcorpan commented on source:
```diff
@@ -7371,8 +7374,229 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+      </ol>
+     </li>
+
+     <li><p>Set [[<var>attr</var>SetViaReflection]] to false.</p></li>
```
Will it not always be false here, given the previous step?

alice replied:
Indeed, and I think I see a problem with that logic as well. As written, you can override the set-by-reflection value just by calling `setAttribute()` twice, which seems silly. Will fix.

zcorpan replied:
If I understand correctly, the point of this thing is to avoid doing work twice when setting the IDL attribute. It will be false after having set the IDL attribute, since the setter sets it to true, but then changes the content attribute, and the attribute change steps then changes it to false. Which seems OK?

But if so, I think there's a bug in the `FrozenArray` setter, since it sets it to false in step 3, and then changes the content attribute in step 7...

alice replied:
I'm trying to get my head around the logic again. I do recall trying to avoid duplicating work (or getting into an infinite loop) but I can't remember exactly how it worked...

zcorpan commented on source:
```diff
@@ -7371,8 +7374,229 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+      </ol>
+     </li>
+
+     <li><p>Set [[<var>attr</var>SetViaReflection]] to false.</p></li>
```
Remove this step (I think)

zcorpan commented:
@annevk @alice thinking a bit more about this, I think the purpose of [[*attr*SetViaReflection]] is to avoid running the attribute change steps from the IDL attribute setter. But wouldn't it be better to be able to set a flag "skip attribute change steps" when changing/removing an attribute, and have the DOM spec deal with that?

annevk commented on source:
```diff
@@ -7371,8 +7374,229 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+      <p>If the given value is:</p>
+      <ul class="brief">
+       <li>null, or</li>
+       <li>not type-compatible with the IDL attribute, or</li>
```
Can this happen? IDL would throw, no?

domenic replied:
I think you're right. This could not be reached, and should be removed. We should add tests that it just throws without resetting any internal state.

annevk commented on source:
```diff
@@ -7371,8 +7374,229 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+
+   <li>
+    <p>The following <span data-x="concept-element-attributes-change-ext">attribute change
+    steps</span> are used to synchronize between the content attribute and the IDL attribute:</p>
```
I think this bit makes it unlike other reflecting attributes. Perhaps this is better introduced as a separate macro-mechanism. E.g., element _x_ supports element-pointer attribute _y_ or some such. As reflecting IDL attributes generally modify the content attribute and the behavior is then defined in terms of those. Though I guess @domenic saw this differently so maybe it's okay?

alice replied:
I have reworked this a bit now and the attribute change steps are much more minimal. The same objection may well still apply, though.

alice commented:
I did some re-working here; @dot-miniscule has been working on exploring implementation for this in Blink, and she uncovered one major flaw with the previous wording: during parsing, the element referred to by an IDREF attribute may not exist, so doing all the work in the setter is actually impractical.

I reworked it to be based on a _conceptual_ relationship, which is resolved on demand, rather than relying on the `[[attrAssociatedElement]]` slot for everything.

I also removed the `[[attrSetViaReflection]]` slot, since it was so confusing and seemed to over-specify what is really quite a simple behaviour, and added an instruction to skip the attribute change steps where necessary. Please let me know if that's not precise enough, or if there's a better way to specify this.

domenic commented on source:
```diff
@@ -7401,8 +7404,222 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+  that inherits from <code data-x=""><code>Element</code></code>, then:
+  <ul>
+   <li><p>The attribute's element can be associated with a particular other element, referred to as
+   its <dfn><var>attr</var>-associated element</dfn>.</p></li>
```
Having both "attr-associated element" and "[[attrAssociatedElement]]", which mean different things, is pretty confusing to me.

It seems that the definition of "attr-associated element" is not referenced anywhere except in the definition of the getter. Could we just instead use that same algorithm to figure out the getter's return value?

Or, maybe the plan is to refer to the "attr-associated element" in other specs, e.g. ARIA? That makes a lot of sense now that I say it. If so, then it'd be good to rename one of the two. And also, it'd be good to have a numbered-steps algorithm for getting the attr-associated element, instead of the current style.

alice replied:
Yeah, it would be nice to be able to refer to the concept elsewhere, exactly.

So, ok, I renamed the slots to `[[attrElement]]`/`[[attrElements]]`, which is a bit icky given the reflected ARIA properties will use the same naming scheme, but I couldn't really think of a better alternative.

I tried to write a numbered-steps algorithm, although it gets a bit weird at the end.

domenic commented on source:
```diff
@@ -7401,8 +7404,222 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+     <p>then set <var>id</var> to the element's <span data-x="concept-ID">ID</span>.
+
+     <li><p>Set the content attribute's value to <var>id</var>, skipping the <span
+     data-x="concept-element-attributes-change-ext">attribute change steps</span>.</p></li>
```
I think you want to skip these specific attribute change steps, not all the possible attribute change steps. Right? If so, a better way would be to have an explicit boolean slot (e.g. [[attrBeingSetViaReflection]]) which you toggle to true, then set the value, then unset the boolean. And then in the attribute change steps you can check that boolean.

Alternately, I think it would work to move step 2 to after step 5. Step 5 would call the attribute change steps, setting [[attrAssociatedElement]] to null. Then step 2 (now step 6) would set it to the given value.

alice replied:
Moved step 2 after step 5 - so simple! Since the attribute change steps are now minimal, it's trivial just to run them each time.

domenic commented on source:
```diff
@@ -7401,8 +7404,222 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+       <li>
+        <p>If <var>element</var> is:</p>
+        <ul>
+         <li><p>null, or</p></li>
```
This cannot happen, since it is `sequence<T>?`. It would have to be `sequence<T?>` for this to be possible.

alice replied:
Removed.

domenic commented on source:
```diff
@@ -7401,8 +7404,222 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+        <p>If <var>element</var> is:</p>
+        <ul>
+         <li><p>null, or</p></li>
+         <li><p>not type-compatible with the IDL attribute's inner type, or</p></li>
```
This cannot happen; the IDL layer will prevent it.

alice replied:
Removed.

domenic commented on source:
```diff
@@ -7401,8 +7404,222 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+       </li>
+
+       <li><p>Otherwise, let <var>id</var> be the element's <code
+       data-x="concept-ID">ID</code>.</p></li>
```
Should be `<span`, not `<code`

alice replied:
Fixed.

domenic commented:
OK, I pushed some editorial work, plus one slight normative change (noted in the commit message) for the frozenarray case. I also properly handled the caching for the frozenarray case, which is annoying (and ideally the IDL spec should make that easier for us, but it doesn't).

One remaining normative question I have is why we have the test

> If element is not a descendant of any of this element's shadow-including ancestors, then continue.

for the frozenarray reflection, but not the single-element reflection. It also seems perhaps redundant with the same-tree (now same-root) checks.

alice commented:
Thanks for the edits! Took me a while to get my head around the change to the `FrozenArray` behaviour, but I got there.

> One remaining normative question I have is why we have the test
> 
> > If element is not a descendant of any of this element's shadow-including ancestors, then continue.
> 
> for the frozenarray reflection, but not the single-element reflection. It also seems perhaps redundant with the same-tree (now same-root) checks.

Here's what I see for single-element reflection:

> On setting, the IDL attribute must perform the following steps:
> - If the given value is not a descendant of any of this element's shadow-including ancestors, then:
>    1. Set this element's explicitly set attr-element to null.
>    2. Remove the content attribute from this element.
>    3. Return.

vs for `FrozenArray`:

> On setting, the IDL attribute must perform the following steps: (...)
> - For each element in the given value:
>    1. If element is not a descendant of any of this element's shadow-including ancestors, then continue.

These are supposed to be analogous, and to avoid being able to set any element reference which is not "visible" to the attribute's element due to Shadow DOM boundaries. The idea is that you can "see out" of Shadow DOM, but not "see into" it.

The same-tree/same-root checks are for reflecting the ID as a string. You can set an element reference which is not in the same tree (but in a tree with a shared ancestor), but it won't reflect to the content attribute.

alice commented:
@annevk One issue we're struggling with is that we don't have any straightforward way of enforcing shadow root boundaries when someone moves an element into a deeper shadow root.

e.g.

```html
<label id="label">Label for input</label>
<custom-combobox>
  #shadow-root (open)
  | <custom-input id="input">
  |   #shadow-root (open)
  |   | <div id="div" contenteditable></div>
  | </custom-input>
</custom-combobox>
```

Say we do:

```js
input.ariaLabelledByElements = [label];
```

That's allowed. And we're not allowed to do this (regardless of shadow root mode):

```js
input.ariaActiveDescendant = "div";
```

However, we can legally do this:

```js
input.shadowRoot.appendChild(label);
```

That would be a weird thing to do, for sure, but there is nothing stopping us as far as I can tell.

My first thought at how to handle this would be to continue to allow `label` to be in `input`'s `aria-labelledby`-associated elements, but not to allow it to be returned by the IDL attribute getter, i.e.

```js
console.log(input.ariaLabelledByElements);   // logs []
```

This means that we're going to have to do the scope check for both setting *and* getting the IDL attributes.

Does this seem reasonable to you? Do you have any other suggestions, or have I missed something?

rniwa commented:
Is it expected / part of this feature that elements continue to be associated even when it's removed from the current tree?? That seems rather unexpected. When an element with a given ID gets removed from a tree, ID ref usually stops working.

alice commented:
> Is it expected / part of this feature that elements continue to be associated even when it's removed from the current tree?? That seems rather unexpected. When an element with a given ID gets removed from a tree, ID ref usually stops working.

It's expected that you can refer to elements in "lighter" shadow trees, yes.

It's a consequence of the way Shadow DOM works that there's no effective way to prevent the behaviour I described when moving nodes, since there is no existing machinery to make things behave differently when moving things between shadow trees than moving within the same tree.

rniwa commented:
> > Is it expected / part of this feature that elements continue to be associated even when it's removed from the current tree?? That seems rather unexpected. When an element with a given ID gets removed from a tree, ID ref usually stops working.
>
> It's expected that you can refer to elements in "lighter" shadow trees, yes (i.e. where IDREFs would fail to be associated).

That's not what I'm talking about. I'm talking about *removing*. When an element is removed from the document (e.g. via `removeChild`), is the element expected to continue to be associated with whatever element it got associated with id-ref arrays?

alice commented:
> That's not what I'm talking about. I'm talking about _removing_. When an element is removed from the document (e.g. via `removeChild`), is the element expected to continue to be associated with whatever element it got associated with id-ref arrays?

Hmm, I must have misunderstood then - I take it this is not in response to my previous comment (https://github.com/whatwg/html/pull/3917#issuecomment-527263562) - or at least tangential.

That is a good question. I think as specced right now, it probably would continue to be associated. 

I can see two options:
1. We could specify that `attr`-associated elements are only to be treated as valid (e.g. for `aria-labelledby`, don't include detached elements when computing the name for an element) when they are in the same document. 
2. Or, we could specify that when a node is removed from the document, some clean-up has to occur to remove it from any `attr`-associated element references as well. This would mean keeping track of the relationship in both directions.

rniwa commented:
> > That's not what I'm talking about. I'm talking about _removing_. When an element is removed from the document (e.g. via `removeChild`), is the element expected to continue to be associated with whatever element it got associated with id-ref arrays?
> 
> Hmm, I must have misunderstood then - I take it this is not in response to my previous comment ([#3917 (comment)](https://github.com/whatwg/html/pull/3917#issuecomment-527263562)) - or at least tangential.

It's in response to your comment. Let me explain why.

> I can see two options:
> 
> 1. We could specify that `attr`-associated elements are only to be treated as valid (e.g. for `aria-labelledby`, don't include detached elements when computing the name for an element) when they are in the same document.
> 2. Or, we could specify that when a node is removed from the document, some clean-up has to occur to remove it from any `attr`-associated element references as well. This would mean keeping track of the relationship in both directions.

If we did (2), then it's moving an element from one place to another would have disassociated it (because in order to insert an element somewhere would require first removing it).

I could see the argument either way. (2) would incur an extra runtime cost during DOM tree mutation like live node lists, or some other kind of house keeping like bidirectional mapping as you suggested.

alice commented:
> If we did (2), then it's moving an element from one place to another would have disassociated it (because in order to insert an element somewhere would require first removing it).

Hm, is there no difference between moving an element and removing it from the document? Would moving within the same scope also cause a disassociation?

rniwa commented:
> > If we did (2), then it's moving an element from one place to another would have disassociated it (because in order to insert an element somewhere would require first removing it).
> 
> Hm, is there no difference between moving an element and removing it from the document? Would moving within the same scope also cause a disassociation?

There is no distinction between moving an element vs. removing & inserting it elsewhere because "moving" an element is defined as removing it from its old parent and inserting under a new parent.

alice commented:
> There is no distinction between moving an element vs. removing & inserting it elsewhere because "moving" an element is defined as removing it from its old parent and inserting under a new parent.

Would you expect moving an element within the same scope to cause it to drop off element reference properties?

rniwa commented:
> > There is no distinction between moving an element vs. removing & inserting it elsewhere because "moving" an element is defined as removing it from its old parent and inserting under a new parent.
> 
> Would you expect moving an element within the same scope to cause it to drop off element reference properties?
> 
> e.g.
> 
> ```
> <input id="input">
> <optgroup>
>   <option id="first">First option</option>
>   <option id="second">Second option</option>
> </optgroup>
> ```
> 
> ```js
> input.ariaActiveDescendantElement = first;
> first.parentElement.appendChild(first);  // move to end of list
> console.log(input.ariaActiveDescendantElement);  // now null?
> ```

That would be the behavior of other normal DOM API.

e.g.
```js
document.body.innerHTML = '<input type="checkbox" form="foo"><form id="foo">';
document.querySelector('input').form; // This would be not null.
document.querySelector('form').remove();
document.querySelector('input').form; // This would be null.
```

If we don't disassociate an element upon removal, we'd get into very strange situations like an input element associated with a form element in a different document, or a single label element associated with multiple input elements in different documents or shadow trees.

alice commented:
Hm, but what if you just moved the form in the DOM? Same result?

rniwa commented:
That code would remove the association and then inserts it back into the DOM, and newly associates it with the form again. That’s one benefit of ID ref referring to whatever element which happens to have the matching string. An explicit association with JS API can’t work like that.

alice commented:
I'm still interested in what you would _expect_ in this case:

```html
<input id="input">
<optgroup>
  <option id="first">First option</option>
  <option id="second">Second option</option>
</optgroup>
```

```js
input.ariaActiveDescendantElement = first;
first.parentElement.appendChild(first);  // move to end of list
console.log(input.ariaActiveDescendantElement);  // ?
```

If we went with option 1 in https://github.com/whatwg/html/pull/3917#issuecomment-528642778, it would still be `first`. This would be analogous to using a globally-unique ID. I think this behaviour would be less surprising, although it would come at the cost of potential memory leaks under limited circumstances (i.e. authors setting a relationship, then removing the associated element from the DOM without also explicitly re-setting the relationship).

If we went with option 2, it would be `null`. This would probably be an unpleasant surprise, but would avoid surprise memory leaks.

rniwa commented:
> I'm still interested in what you would _expect_ in this case:
> 
> ```
> <input id="input">
> <optgroup>
>   <option id="first">First option</option>
>   <option id="second">Second option</option>
> </optgroup>
> ```
> 
> ```js
> input.ariaActiveDescendantElement = first;
> first.parentElement.appendChild(first);  // move to end of list
> console.log(input.ariaActiveDescendantElement);  // ?
> ```
> 
> If we went with option 1 in [#3917 (comment)](https://github.com/whatwg/html/pull/3917#issuecomment-528642778), it would still be `first`.

I'd expect `input.ariaActiveDescendantElement` to be null in that case.

> This would be analogous to using a globally-unique ID.

It's not. ID string simply matches whatever element which happens to be the first element with the same ID string. The way I think of ID ref is that it's referring to an abstract notion of any element with a given ID which appears first in the tree order. I'm not associating it with any particular instance of an element.

With JS API, I'm associating a particular instance of an element, not any abstract notion or relation which ends up resolving to an element as with the case with string based ID referencing.

alice commented:
> I'd expect `input.ariaActiveDescendantElement` to be null in that case.

Could you expand on why this shouldn't be surprising? This would be inconsistent with an element reference in script, which persists regardless of whether an element is moved in the DOM (or even removed).

> With JS API, I'm associating a particular instance of an element, not any abstract notion or relation which ends up resolving to an element as with the case with string based ID referencing.

I agree that that is a difference in practice; I would argue that ID based referencing is _intended_ to be used this way, though (i.e. associating a specific element) - it's an [authoring error](https://html.spec.whatwg.org/multipage/dom.html#global-attributes:tree) ("the id attribute value must be unique amongst all the IDs in the element's tree") have multiple elements with the same ID.

rniwa commented:
> > I'd expect `input.ariaActiveDescendantElement` to be null in that case.
> 
> Could you expand on why this shouldn't be surprising? This would be inconsistent with an element reference in script, which persists regardless of whether an element is moved in the DOM (or even removed).

Because the element got removed? I don't think arguing one or another based on one's opinions is productive here. We need an objective evidence to suggest either behavior is better / more desirable.

> > With JS API, I'm associating a particular instance of an element, not any abstract notion or relation which ends up resolving to an element as with the case with string based ID referencing.
> 
> I agree that that is a difference in practice; I would argue that ID based referencing is _intended_ to be used this way

I disagree.

rniwa commented:
Ugh... after today’s discussion, I noticed a new issue with this current proposal. We can’t really do “invalidation” either during insertion, removal, or on get if it’s a frozen array since we can’t mutate an array without potentially executing arbitrary scripts.

So these properties probably need to be a new subclass of NodeList which auto-updates, or these IDL attributes need to be replaced by getter/setter pairs where getting would return a new array each time.

domenic commented:
Can you explain why changing the value returned by the property could run arbitrary scripts?

rniwa commented:
> Can you explain why changing the value returned by the property could run arbitrary scripts?

Sorry, I was getting confused there. Now that I about it more carefully, this isn’t an issue because it’s not like we’re going to keep the array object assigned by the author in the setter.

alice commented:
@rniwa Would you mind taking a look at the updated spec language here to see if it matches your understanding of what we discussed in the AOM meetings? Thank you!

rniwa commented:
Hm... I thought we were going to remove the reference when a node is adopted into another document? Where is that done?

alice commented:
We decided that probably wasn't worth the extra complexity.

Is there some specific reason we would want to clear the content attribute in that case?

rniwa commented:
> Edit for clarification: the explicitly-set _attr_ element reference is only retained if there is another reference to it, and it can't be used while the element is in another document. In the extremely unlikely case that the element was moved _back_ into its original document, the reference would begin working again.

Oh, right. Now I remember that discussion. Sounds good to me.

alice commented on source:
```diff
@@ -7319,10 +7319,6 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
   <var>attr</var>-elements</dfn>, which is either a <span>list</span> of elements or null. It is
   initially null.</p></li>
 
-  <li><p>Elements of the type this IDL attribute appears on have <dfn>cached
```
@domenic @annevk - @chrisosaurus and I were looking into implementing this, and realised that since this is now being effectively re-computed on each get (to check validity), there's not really anything gained from the caching layer other than the ability to assert that `x.ariaLabelledByElements === x.ariaLabelledByElements`, and it seems like a lot of machinery to set up for a very obscure gain.

In particular, since the following was never true:

```js
let foo = [a, b, c];
bar.ariaActiveDescendantElements = foo;
bar.ariaActiveDescendantElements === foo;  // false with or without caching
```

... it didn't seem worthwhile to keep the caching layer around.

Specifically, we found that to implement this in Blink, we would effectively have to keep _three_ copies of the array:
1) The explicitly-set attr-elements
2) A copy of the attr-associated elements in Blink, to check for staleness, and
3) A copy of the attr-associated elements in V8, to return to the getter if the Blink-side copy isn't stale.

What do you think?

domenic replied:
> other than the ability to assert that x.ariaLabelledByElements === x.ariaLabelledByElements, and it seems like a lot of machinery to set up for a very obscure gain.

This is really important. See https://w3ctag.github.io/design-principles/#attributes-like-data . So IMO it's worth it.

alice replied:
Ah, good point. 

I've reverted this change; triple-caching here we come.

chrisosaurus replied:
@domenic I'm a little concerned about callers relying upon array equality (either with or without a cache)

Without a cache the behavior is more consistent, array equality never works.

With a cache the behavior seems inconsistent - as array equality now *sometimes* works - but can be broken easily, which may lead to subtle errors.

I've tried to put together an example where the presence of a console.log statement reading the property changes the equality outcome
https://gist.github.com/chrisosaurus/c3e4c033184e27990cc03bf5abe4f452#observing-cache-invalidation

I don't want to delay this further, so if this behavior is more tolerable than the cache-less equivalent then so be it, I just wanted to make sure we committed to the ramifications.

domenic replied:
If gets flush the cache, per your

```js
console.log(fruit.ariaDescribedByElements); // this get flushes the cache.  <=====
```

then something is broken in the algorithms. I haven't looked at the spec in detail, but gets should not be side-effecting in this way. I agree this is quite bad.

We should, per the design principles linked above, be able to guarantee both: (a) `x.ariaLabelledByElements === x.ariaLabelledByElements`; and (b) gets are not side effecting.

alice replied:
The gets are inherently side effecting because we need to re-compute the value each time we do a get (even if we return a previously returned value), because we check for validity on get. In Chris' example, the value changes in between each get, so each get returns a different value; the first and third get return equivalent (but not `===`) values.

I'm not sure there's a way we can ensure both (a) and (b) *and* check for validity on each get.

domenic replied:
I guess checking for validity on get is a requirement from some other parties? :(

rniwa replied:
We need to check on get to avoid having relationships between nodes in different shadow trees & documents.

alice replied:
Checking for validity also means we have a closer relationship between the value returned by the getter and what AT APIs will see.

domenic replied:
Ok. It sounds to me like these aren't acting like normal HTML reflection at all anymore, so at this point I'd suggest moving to get/set methods like I suggested above. (Then you can remove the triple caching.)

alice replied:
Can you remind me why it's preferable to have `FrozenArray<Element>` vs. a live result like `HTMLFormControlsCollection`?

domenic replied:
That would be another direction we could go; we'd want to use https://heycam.github.io/webidl/#idl-observable-array to get liveness though (to avoid the web-developer facing pain of fake arrays that are missing all their usual methods, fail `Array.isArray`, etc.). Unfortunately nobody has yet implemented observable array infrastructure in their bindings (to my knowledge), and I'd be hesitant to block this already long-developed feature on such work...

alice replied:
It seems like we're blocked on either that, or re-working the entire spec to individually define setters and getters for each of these attributes :\

domenic replied:
Well, I think you could define "generated" get/set methods in the same way the current draft defines "generated" getters/setters backing a single IDL attribute.

alice replied:
Hmmm the current spec refers to the concept of reflection to IDL attributes; is there any precedent for reflection to explicit setter/getter methods?

domenic replied:
No, but there's also no precedent for reflecting elements/arrays of elements :). So I think it's similar amounts of novelty either way, with the unifying concept being "reflection means generating IDL that automatically retrieves/manipulates content attributes".

alice commented on source:
```diff
@@ -7378,9 +7379,6 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
     </li>
 
-    <li><p>If <var>elements</var> <span data-x="list is empty">is empty</span>, then set
-    <var>elements</var> to null.</p></li>
```
@domenic I was going through the WPT tests and this behaviour seemed wrong to me, so I changed it: 

Previously, if there were no explicitly set _attr_-elements and the content attribute didn't yield any valid elements, we returned null.

Instead, I've changed it so that if the content attribute isn't present, it returns null, but otherwise if the content attribute is present, there are no explicitly set _attr_-elements and the content attribute doesn't yield any valid elements, it returns an empty array.

This seems more symmetrical to me with being able to set the IDL attribute to null to clear the content attribute.

domenic replied:
That makes perfect sense to me; thanks for the update and the concise summary!

rniwa commented:
Hm... do we want to remove duplicates from FrozenArray upon assignment (setter) here?

alice commented:
> Hm... do we want to remove duplicates from FrozenArray upon assignment (setter) here?

I'm not sure I follow. IDREF-list attributes don't disallow duplicate values as far as I know; unless there's a specific reason to disallow them here, I don't understand why we would add that constraint. (I also can't see any reason why authors would want to be able to use duplicate values, but I'm not that imaginative.)

rniwa commented:
> > Hm... do we want to remove duplicates from FrozenArray upon assignment (setter) here?
> 
> I'm not sure I follow. IDREF-list attributes don't disallow duplicate values as far as I know; unless there's a specific reason to disallow them here, I don't understand why we would add that constraint. (I also can't see any reason why authors would want to be able to use duplicate values, but I'm not that imaginative.)

`DOMTokenList` de-duplicates the entires. e.g.
```
span = document.createElement('span');
span.className = 'foo bar foo'
span.classList.length // 2
span.classList[0] // foo
span.classList[1] // bar
span.classList.value // foo bar foo
```

alice commented:
> > > Hm... do we want to remove duplicates from FrozenArray upon assignment (setter) here?
> > 
> > 
> > I'm not sure I follow. IDREF-list attributes don't disallow duplicate values as far as I know; unless there's a specific reason to disallow them here, I don't understand why we would add that constraint. (I also can't see any reason why authors would want to be able to use duplicate values, but I'm not that imaginative.)
> 
> `DOMTokenList` de-duplicates the entires. e.g.
> 
> ```
> span = document.createElement('span');
> span.className = 'foo bar foo'
> span.classList.length // 2
> span.classList[0] // foo
> span.classList[1] // bar
> span.classList.value // foo bar foo
> ```

Thanks for that!

I think in that case de-duplicating makes sense, because the classlist is really more of a class set - adding another class, or changing the order, wouldn't make a difference.

With an IDREF list, the order is significant, and adding a duplicate entry would change the behaviour.

For example, if I had something like `aria-labelelledby="foo foo"`, that should cause the contents of the `"foo"` element to be included twice in the name of the element with the `aria-labelledby` attribute on it.

rniwa commented:
> > `DOMTokenList` de-duplicates the entires. e.g.
> > ```
> > span = document.createElement('span');
> > span.className = 'foo bar foo'
> > span.classList.length // 2
> > span.classList[0] // foo
> > span.classList[1] // bar
> > span.classList.value // foo bar foo
> > ```
> 
> I think in that case de-duplicating makes sense, because the classlist is really more of a class set - adding another _instance of the same_ class, or changing the order, wouldn't make a difference.
> 
> With an IDREF list, the order is significant, and adding a duplicate entry would change the behaviour.
> 
> For example, if I had something like `aria-labelelledby="foo foo"`, that should cause the contents of the `"foo"` element to be included twice in the name of the element with the `aria-labelledby` attribute on it.

Hm... that's interesting. In the case of [slot assignment](https://github.com/whatwg/html/issues/3534), we need to remove duplicates but that doesn't really have a setter so it works fine there, and I guess there aren't really any other DOM API that uses ID ref list than ARIA?

annevk commented:
There is: https://html.spec.whatwg.org/#attr-tdth-headers. That's currently defined to take an unordered set and there's no API for it that would expose order or some such. I guess this would be the first ordered list of IDs. (Note that `DOMTokenList` is ordered, but it's indeed a set despite the name, not a list.)

mrego commented:
This PR has been on hold for more than a year now, so let's do an update.

Chromium and WebKit have now an implementation of this PR under a runtime flag (`AOMAriaRelationshipProperties` flag in Chromium and `AriaReflectionForElementReferencesEnabled` flag in WebKit).
Both implementations allow to refer to descendants of your shadow-including ancestors (https://github.com/whatwg/html/issues/6063).

There's a WPT test covering this PR too: https://github.com/web-platform-tests/wpt/blob/master/dom/nodes/aria-element-reflection.tentative.html (Chromium [fails one case](https://bugs.chromium.org/p/chromium/issues/detail?id=1321564) but that's being currently fixed).

Does anyone remember what's blocking this PR? What else should be done before being ready to merge? Thanks!

annevk commented:
I think a fresh (rebased) PR with a filled out PR template (that wasn't part of our setup back in 2018) would be good at this point, properly acknowledging @alice for her contributions toward this feature.

mrego commented:
Thanks @annevk, I've just done a rebase and created a new PR: https://github.com/whatwg/html/pull/7934
If I should have updated this PR instead, please let me know how.

mrego commented:
This has been moved to https://github.com/whatwg/html/pull/7934, which is about to land; so I guess we can close this one now.
